### ISSUE TITLE
Unable to Open Link "URL was blocked by a content blocker"

### ISSUE THREAD

**Issue by evancrow** [neevaco/neeva-ios#3779]
I've noticed sometimes when navigating a URL (especially package tracker links) that I can't see the site due to a content blocker. Safari has a way to override this, think it would make sense to add that option.

<img src=https://user-images.githubusercontent.com/37356984/172069450-757ebc66-10d9-400c-806c-6d310ff64a3d.jpeg width=300>

Also, I noticed the warning isn't centered on the page...

FYI @chung-lai

**Comment by darinwf** [comment]
The option in the shield menu dropdown to try disabling cookie cutter should be the way we let users override cookie cutter for just this site. Maybe we just need to hook that up?

**Comment by evancrow** [comment]
Maybe? I tried disabling Cookie Cutter and all the tracking protection but the website still won't load.

**Comment by chung-lai** [comment]
interesting, we should add the domain to the Safari list and Safari should take care of it. 🤔  Will take a look

**Comment by chung-lai** [comment]
@reneehuang1 reported a promo is being blocked as well:

![Image from iOS (6)](https://user-images.githubusercontent.com/87332578/172305348-b972c7b7-036c-4136-b551-902c00819d80.png)

**Comment by chung-lai** [comment]
When I look at easylist vs easyprivacy. I start to wonder if strict mode is doing the right thing when we block request with EasyPrivacy as I found `uniqlo` in the easyPrivacy rules.

**Comment by darinwf** [comment]
Or perhaps what we are missing is that these filters should only be applied to sub-resource requests and not top-level frame requests? Is there a way to specify that on the WebKit APIs?

**Comment by darinwf** [comment]
Also can we repro these issues with Firefox when strict mode is enabled there?

**Comment by chung-lai** [comment]
> Or perhaps what we are missing is that these filters should only be applied to sub-resource requests and not top-level frame requests? Is there a way to specify that on the WebKit APIs?

There are a couple trigger parameters the [WebKit API](https://developer.apple.com/documentation/safariservices/creating_a_content_blocker) provides:

```
if-top-url | An array of strings matched to the entire main document URL; limits the action to a specific list of URL patterns. Values must be lowercase ASCII, or Punycode for non-ASCII. Can't be used with unless-top-url.
-- | --
unless-top-url | An array of strings matched to the entire main document URL; acts on any site except URL patterns in provided list. Values must be lowercase ASCII, or Punycode for non-ASCII. Can't be used with if-top-url.
```

**Comment by chung-lai** [comment]
> Also can we repro these issues with Firefox when strict mode is enabled there?

Good point, doesn't seem to be an issue with Firefox (at least for the uniqlo promo). Will check out the list and see if there is anything missing from our list.

**Comment by darinwf** [comment]
> > Also can we repro these issues with Firefox when strict mode is enabled there?
> 
> Good point, doesn't seem to be an issue with Firefox (at least for the uniqlo promo). Will check out the list and see if there is anything missing from our list.

Cool, hopefully by studying the differences between us and Firefox, we can see what we are missing.

`unless-top-url` sounds like the mechanism we should be using to disable content blocking on a particular site?

**Comment by evancrow** [comment]
@chung-lai The problem is persisted even with tracking protection off, so does that mean it's not the list rules? Or is the toggle not working right and it's two separate issues?

**Comment by chung-lai** [comment]
When I completely turn off cookie cutter toggle in settings, the uniqlo promo page loads for me. Is that the toggle you are talking about?

**Comment by chung-lai** [comment]
Yes, i think that's a separate issue. For me, it's off when I tried to load those blocked pages and that toggle doesn't seem to work for them.

**Comment by chung-lai** [comment]
@evancrow what's the url you are having problems? Could you please post them here if it nothing personal?

**Comment by chung-lai** [comment]
So looks like it's the `mi_ecmp` that prevents the uniqlo promo from loading:

```
http://t.usa.uniqlo.com/r/?id=he160b74d,3ad4bcdb,28895ad8&utm_medium=email&utm_campaign=20220606-SailorMoonUT&bxid=3255259&recipientID=3255259&broadlogID=-513755315&utm_source=NL-Ad-Hoc&mi_u=3255259&mi_ecmp=-513755315
```

and the rule:

```
    {
        "action": {
            "type": "block"
        },
        "trigger": {
            "url-filter": "^https?://.*&mi_ecmp="
        }
    },
```

**Comment by evancrow** [comment]
I lost my link but a family member who pointed me to the issue sent me this: https://eliotsnutbutters.us7.list-manage.com/track/click?u=e538eb600fa31aa6c156c6045&id=9e0dcf62ec&e=8f705ececd

**Comment by chung-lai** [comment]
Regarding the toggle problem: for those blocked pages, we show the internal error page which sets the tab url to something like `internal://localerror...url=http://.....`. When you toggle, it will add `local` instead of the actual domain to the allow list.

However, even I tried to add those to the allowlist, it only works for eilotsnutbutters but not uniqlo promo. 🤔

**Comment by evancrow** [comment]
@chung-lai Another report [here](https://github.com/neevaco/neeva/issues/67245)

**Comment by chung-lai** [comment]
After studying a bit more with firefox rules, look like most of their rules have this:

```
"load-type":["third-party"]
```

And the documentation for this attribute from apple documentation:

```
load-type:
An array of strings that can include one of two mutually exclusive values. If not specified, the rule matches all load types. first-party is triggered only if the resource has the same scheme, domain, and port as the main page resource. third-party is triggered if the resource isn’t from the same domain as the main page resource.
```

**Comment by chung-lai** [comment]
Now I start to wonder if this is what we should do:

* Add this `["third-party"]` to all the rules in the strict rule list so it doesn't block any request that's coming from the origin domain
* For strict mode, we will load two lists: 
  * the standard cookie blocking list so it should block all the cookies on the same page
  * the new strict easy-privacy list which will have `["third-party"]` added to all the rules

With this, it should be sufficient to block all cookie settings on the page which should prevent tracking and it will block all the third party request on the page itself.

@darinwf @evancrow, wdyt? We can chat more offline.

**Comment by evancrow** [comment]
@chung-lai That sounds reasonable to me. I don't know the tracking rule stuff very well though so I don't have much input.

**Comment by darinwf** [comment]
I think it is worth a try adding `third-party`.. it might make our blocker too lax in some cases, but it should workaround this issue which is a current frustration for users

**Comment by evancrow** [comment]
Just a thought, but could we detect if the URL is being blocked and then for that site only, adjust the rules to standard instead of strict?

**Comment by chung-lai** [comment]
that's an interesting idea, switching between the two does have performance implication but that's doable. Only thing is the page has to be loaded before we can detect it, we may have to add an intermediate step for page load. It's doable though. Will start with the `third-party` approach to see if it fixes most of the issues. Will continue digging more on how we use the easyPrivacy list and explore other ideas.
